realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -0,0 +1,203 @@
+                                                      "([Ljava/lang/String;)V");
+            env->CallVoidMethod(row_obj, notify_change_listeners, m_field_names_array);
+        });
+    }
+
+    void error(std::exception_ptr err)
+    {
+        if (err) {
+            try {
+                std::rethrow_exception(err);
+            }
+            catch (const std::exception& e) {
+                Log::e("Caught exception in object change callback %1", e.what());
+            }
+        }
+    }
+
+private:
+    ObjectWrapper* m_wrapper;
+    bool m_deleted = false;
+    jobjectArray m_field_names_array = nullptr;
+};
+
+static void finalize_object(jlong ptr)
+{
+    TR_ENTER_PTR(ptr);
+    delete reinterpret_cast<ObjectWrapper*>(ptr);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeGetFinalizerPtr(JNIEnv*, jclass)
+{
+    TR_ENTER()
+    return reinterpret_cast<jlong>(&finalize_object);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeCreate(JNIEnv*, jclass, jlong shared_realm_ptr,
+                                                                      jlong row_ptr)
+{
+    TR_ENTER_PTR(row_ptr)
+
+    // FIXME: Currently OsObject is only used for object notifications. Since the Object Store's schema has not been
+    // fully integrated with realm-java, we pass a dummy ObjectSchema to create Object.
+    static const ObjectSchema dummy_object_schema;
+
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    auto& row = *(reinterpret_cast<Row*>(row_ptr));
+    Object object(shared_realm, dummy_object_schema, row); // no throw
+    auto wrapper = new ObjectWrapper(object);              // no throw
+
+    return reinterpret_cast<jlong>(wrapper);
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsObject_nativeStartListening(JNIEnv* env, jobject instance,
+                                                                             jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    try {
+        auto wrapper = reinterpret_cast<ObjectWrapper*>(native_ptr);
+        if (!wrapper->m_row_object_weak_ref) {
+            wrapper->m_row_object_weak_ref = JavaGlobalWeakRef(env, instance);
+        }
+
+        // The wrapper pointer will be used in the callback. But it should never become an invalid pointer when the
+        // notification block gets called. This should be guaranteed by the Object Store that after the notification
+        // token is destroyed, the block shouldn't be called.
+        wrapper->m_notification_token = wrapper->m_object.add_notification_block(ChangeCallback(wrapper));
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsObject_nativeStopListening(JNIEnv* env, jobject, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    try {
+        auto wrapper = reinterpret_cast<ObjectWrapper*>(native_ptr);
+        wrapper->m_notification_token = {};
```

Comment:
`m_notification_token` is an instance but not a pointer. so you cannot give a `nullptr` value to it. `{}` is the efficient way here since it will destruct the old instance and give it empty values. See definition of `NotificationToken& operator=(NotificationToken&&);`